### DISCUSSION TITLE
Update cover status to opening/closing as soon as it accepts a command

### DISCUSSION THREAD

**Discussion by mariusz-schimke** [zwave-js/zwave-js#7176] [Feature Requests]
### Description

I use Fibaro FGR-223 controllers for window blinds and also a Nice IBT4ZWAVE module for controlling a gate. What I noticed from the very beginning of using them is that you don't see the current state of the device until it stops its motion. So you tap an open or close button on the GUI or click a physical button, but in Home Assistant the state doesn't change to opening/closing. Only when you wait until the blind/gate fully opens/closes or when you stop it, does the state update to open/closed. So, for instance, when you tap a button to open/close the gate, you don't know if the tap was accepted as a command or not, and you learn about that only 40 seconds later ;).

I'm mentioning HA here, but I did read your checklist where to report issues. Apologies if you're not the right addressee, but my gut feeling tells me the behavior may rather be dependent on the core driver rather than HA itself. Please correct me if I'm wrong. I initially opened an issue [here](https://github.com/hassio-addons/addon-zwave-js-ui/issues/707), but then realized it may have nothing to do with the HA add-on.

### Do you have a solution in mind?

Initially I thought this was because of how the devices just work. I assumed that they only dispatch events when they finish their motion, and that's why we don't know that they are in transition from one state to another. But I analyzed the logs recently and noticed that as soon as the device receives a command, it confirms it. Even if the command was initiated outside the software, that is by the physical button.

Please have a look a the log files I attached. I added a bit of information at their beginning for reference.
[zwave_js-gate-closing-triggered-by-z-wave-controller.log](https://github.com/user-attachments/files/17045507/zwave_js-gate-closing-triggered-by-z-wave-controller.log)
[zwave_js-gate-opening-triggered-by-an-external-button.log](https://github.com/user-attachments/files/17045508/zwave_js-gate-opening-triggered-by-an-external-button.log)

### Describe alternatives you've considered

I'm wondering if you could make use of the command confirmations the devices respond with, to immediately notify about a state change. This way HA (or other software) could make a transition to opening/closing and update the state right away. I find this useful as sometimes I open/close the gate without seeing it. Then I don't know if it's actually going to open/close, and I have to recheck it after 40 seconds ;).

**Comment by kpine** [comment]
> I assumed that they only dispatch events when they finish their motion, and that's why we don't know that they are in transition from one state to another. But I analyzed the logs recently and noticed that as soon as the device receives a command, it confirms it. Even if the command was initiated outside the software, that is by the physical button.

Just so we're on the same page, which messages are you referring to when you say this?

If you are referring to `targetValue`, HA doesn't use this to report entity state. Is uses `currentValue`.

**Reply by AlCalzone** [reply]
For the first log:
The `setValue` API call returns a Supervision result with status `WORKING`, so it's possible for the HA integration to indicate that the transition is in progress, until the update for `currentValue` is received.

For the second log:
When the device is opened, it sends a report with target value 99, and current value "unknown". Once it is done, it sends another report with identical target value and current value identical.
This could be used to indicate a transition aswell. Although I'm not sure if it is foolproof to assume that `targetValue != currentValue` indicates a transition, since for example a window cover could manually be stopped halfway. I didn't find anything in the specs that the target value has to be updated in this case, only that it **may**:
> The Target Value field MUST advertise the target value of an ongoing transition or the most recent transition.
> [...]
> The Target Value may change at a later time due to local control or a Multilevel Switch Stop Level Change command.

In any case, `node-zwave-js` exposes the relevant functionality, so this is really a feature request for the Z-Wave integration in HA.

**Reply by mariusz-schimke** [reply]
Many thanks for a detailed response! Could you please advise me where I could post an issue regarding this? I posted it [here](https://github.com/hassio-addons/addon-zwave-js-ui/issues/707), but have an impression that this repository is mostly a dockerfile. May be wrong, but it doesn't look like there is any relevant code. I'm not sure where to post it then. Maybe [home-assistant/core](https://github.com/home-assistant/core/issues/new/choose)? Will be grateful for pointing me to the right place ;).

Regarding how the `targetValue` and `currentValue` correspond to each other on stop, I can check that if that may be helpful. Of course I'm not sure if my devices could be used as a the only source of truth, that is if all devices of this class will behave the same.

**Reply by AlCalzone** [reply]
https://github.com/home-assistant/core/issues/new/choose, but that just refers you to the community forums.

> Of course I'm not sure if my devices could be used as a the only source of truth

If I've learned anything over the last few years: no :(

**Reply by mariusz-schimke** [reply]
Thanks! I just posted a request [here](https://community.home-assistant.io/t/update-cover-status-to-opening-closing-while-in-transition-to-open-closed/772447).
